sharkfuser/build_tools/cmake/CTestMacros.cmake

```diff
@@ -82,10 +139,84 @@ function(add_sharkfuser_sample)
     return()
   endif()
 
-  _add_sharkfuser_target(
+  _add_sharkfuser_test(
     NAME ${_RULE_NAME}
     SRCS ${_RULE_SRCS}
     DEPS ${_RULE_DEPS}
     BIN_SUBDIR samples
   )
 endfunction()
+
+
+# Creates a lit test that compiles a source file and runs lit on it.
+#
+#  add_sharkfuser_lit_test(
+#    SRC <source-file>
+#    [TOOLS <tool1> <tool2> ...]
+#    [DEPS <dep1> <dep2> ...]
+#  )
+#
+# SRC
+#  The source file to compile and test (required)
+#
+# TOOLS
+#  External tools needed for the test (e.g., FileCheck)
+#
+# DEPS
+#  Library dependencies for the compiled executable
+function(add_sharkfuser_lit_test)
+  if(NOT SHARKFUSER_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE               # prefix
+    ""                  # options
+    "SRC"               # one value keywords
+    "TOOLS;DEPS"        # multi-value keywords
+    ${ARGN}
+  )
+  if(NOT _RULE_SRC)
+    message(FATAL_ERROR "add_sharkfuser_lit_test: SRC parameter is required")
+  endif()
+
+  get_filename_component(_ABSOLUTE_RULE_SRC ${_RULE_SRC} ABSOLUTE)
+  get_filename_component(_TEST_NAME ${_RULE_SRC} NAME_WE)
+
+  # The executable who's output is being lit tested.
+  _add_sharkfuser_test_target(
+    NAME ${_TEST_NAME}
+    SRCS ${_RULE_SRC}
+    DEPS ${_RULE_DEPS}
+    BIN_SUBDIR samples
+  )
+
+  # Pass lit locations of tools in build directory through `--path` arguments.
+  set(_LIT_PATH_ARGS "--path" "$<TARGET_FILE_DIR:${_TEST_NAME}>")
+  foreach(_TOOL IN LISTS _RULE_TOOLS)
+    list(APPEND _LIT_PATH_ARGS "--path" "$<TARGET_FILE_DIR:${_TOOL}>")
+  endforeach()
+
+  add_test(
+    NAME
+      ${_TEST_NAME}
+    COMMAND
+      "${Python3_EXECUTABLE}"
+      "${LLVM_EXTERNAL_LIT}"
+      "${_ABSOLUTE_RULE_SRC}"
+      ${_LIT_PATH_ARGS}
+      # lit config sets the "%test_exe" substitution based on this param.
+      "--param" "TEST_EXE=$<TARGET_FILE:${_TEST_NAME}>"
+      # Ensures lit prints a (more) useful error message on failure.
+      "--verbose"
+  )
+
+  # Apparently this flag helps FileCheck spit out nicer error messages.
+  set_tests_properties(${_TEST_NAME} PROPERTIES ENVIRONMENT
+    "FILECHECK_OPTS=--enable-var-scope")
+
+  # Dependencies for the test.
+  if(_RULE_TOOLS)
+    set_tests_properties(${_TEST_NAME} PROPERTIES DEPENDS "${_RULE_TOOLS}")
+  endif()
```

Comment:
If not too much trouble, recommend moving these into a separate function to keep `add_sharkfuser_lit_test` easy to follow.

Reply:
Which parts would you recommend moving into a separate function? Too me all of this stuff in this function  hangs together pretty well.

Reply:
I forget why I said this. I think I was trying to keep the same pattern across the user-facing macros, but let's leave it how you have it now.
```

- add_sharkfuser_test: parse cmake args, early return, invoke `_add_sharkfuser_ctest_target` with args
- add_sharkfuser_sample: parse cmake args, early return, invoke `_add_sharkfuser_ctest_target` with args
- add_sharkfuser_lit_test: parse cmake args, early return, invoke `_add_sharkfuser_lit_test_target` with args 

```